pkg/internalregistry/controllers/service_account_controller.go

```diff
@@ -134,16 +139,56 @@ func (c *serviceAccountController) sync(ctx context.Context, key string) error {
 	// name of managed secret
 	secretName, err := c.managedImagePullSecretName(ctx, serviceAccount)
 
-	// ensure secret ref annotation is set
-	if serviceAccount.Annotations[InternalRegistryImagePullSecretRefKey] != secretName {
-		patch, err := applycorev1.ExtractServiceAccount(serviceAccount, serviceAccountControllerFieldManager)
-		if err != nil {
-			return err
+	// extract the apply configuration patch from the cache
+	patch, err := applycorev1.ExtractServiceAccount(serviceAccount, serviceAccountControllerFieldManager)
+	if err != nil {
+		return err
+	}
+
+	// we always want the annotation set
+	patch.WithAnnotations(map[string]string{InternalRegistryImagePullSecretRefKey: secretName})
+
+	// TODO stop adding to .Secrets part of API-1798
+	if !slices.ContainsFunc(patch.Secrets, func(ref applycorev1.ObjectReferenceApplyConfiguration) bool { return *ref.Name == secretName }) {
+		patch.WithSecrets(applycorev1.ObjectReference().WithName(secretName))
+	}
+
+	// We have tp be careful with the atomic ServiceAccount.ImagePullSecrets field so
+	// that we do not race with another controller or user. Add ImagePullSecrets to
```

Comment:
```suggestion
	// that we do not contend with another controller or user over field ownership. Add ImagePullSecrets to
```

Would that be more precise? Basically, this controller expects that some ServiceAccounts will have others writing to the atomic ImagePullSecrets list, but as long as other writers preserve the specific entry it cares about, it won't insist that it is also the field's owner.

Reply:
Reworded.